fs-index/src/tests.rs

```diff
@@ -0,0 +1,699 @@
+use std::{fs, path::Path};
+
+use anyhow::{anyhow, Result};
+use tempfile::TempDir;
+
+use data_resource::ResourceId;
+use dev_hash::Crc32;
```

Comment:
We should have tests using cryptographic hash function as well.

I see two approaches:
1. Copy-paste and adjust, so we'd have 2 files: `tests_crc32.rs` and `tests_blake3.rs`.
2. Write simple automation function/macro and convert existing test cases into automated test cases, where we iterate through vector of hash functions (2 items is enough but could be more) and run the test case itself inside the loop. We could also define 2 vectors: cryptographic and non-cryptographic, some test cases should be written differently, on top of these vectors.

Approach 1 will reduce maintainability of the codebase, while approach 2 is probably overkill and could also decrease readability. Probably, we could have something in between?

Reply:
I added a conditional import statement, so we don't need to duplicate the entire test suite. This way, we also avoid the need for additional automation functions or macros.

I've included running the tests with both hash functions in our CI workflows.

Let me know what you think of b3eeda0f3e52a1055c031e667cf847dc68925385

Reply:
Not bad, but I think with this approach we might run into problems later.

I think we should use approach 2 described above. The macro can be defined in utils, that takes a specific function and runs some computation with this function.

Then our tests could be split into 3 files:
* `tests_all.rs` for testing invariants that must hold regardless of the type of hash function
* `tests_crypto.rs` for testing invariants that hold true only for cryptographic hash functions
* `tests_fast.rs` for testing invariants that hold true only for non-cryptographic functions

So, each unit test would look like:
```
with_hash_function([blake3, crc32, sha256]) {
...
}
```

or separate macros for hash function types:
```
with_crypto_hashing {
...
}

with_fast_hashing {
...
}
```

This way it would be easier to control what functions are used in each test.

Reply:
I agree, this is a better approach. 

My initial suggestion might have been confusing `fs-index` users, as the feature flags don’t add value to the library itself. Instead, I defined a macro `generate_tests_for_hash` to generate tests for a given hash type (implementing `ResourceId`). This macro is used to run tests for both `Blake3` and `CRC32`, and it can be extended for any hash type from the `dev_hash` crate.

The module organization is clear. Here’s how the tests look in the terminal:
```shell
running 22 tests
test tests::blake3::test_build_index_with_file ... ok
test tests::blake3::test_build_index_with_directory ... ok
test tests::blake3::test_build_index_with_empty_file ... ok
test tests::blake3::test_build_index_with_multiple_files ... ok
test tests::blake3::test_hidden_files ... ok
test tests::blake3::test_build_index_with_multiple_directories ... ok
test tests::crc32::test_build_index_with_directory ... ok
test tests::crc32::test_build_index_with_file ... ok
test tests::crc32::test_build_index_with_empty_file ... ok
test tests::blake3::test_add_colliding_files ... ok
test tests::blake3::test_store_and_load_index ... ok
test tests::crc32::test_add_colliding_files ... ok
test tests::blake3::test_resource_index_update ... ok
test tests::blake3::test_num_collisions ... ok
test tests::crc32::test_hidden_files ... ok
test tests::crc32::test_build_index_with_multiple_files ... ok
test tests::blake3::test_store_and_load_index_with_collisions ... ok
test tests::crc32::test_build_index_with_multiple_directories ... ok
test tests::crc32::test_store_and_load_index ... ok
test tests::crc32::test_num_collisions ... ok
test tests::crc32::test_resource_index_update ... ok
test tests::crc32::test_store_and_load_index_with_collisions ... ok

test result: ok. 22 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.02s
```
I chose to keep all test modules in one file for simplicity. Each module for a specific hash function is only a few lines, so it keeps the file concise.

Is this fine or do you think each of the modules should be in a separate file?